File: 0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> mp;

        for(string str : strs) {
            string val = str;
            sort(str.begin(), str.end());
            string key = str;
            mp[key].push_back(val);
        }

        vector<vector<string>> ans;

        for(auto it : mp) {
            ans.push_back(it.second);
        }

        return ans;
    }
};

File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        HashMap<String, List<String>> mp = new HashMap<>();

        for(String str : strs) {
            //pre-store the untampered val
            String val = str;
            
            //convert to char array
            char[] charArr = str.toCharArray();

            //now sort the array
            Arrays.sort(charArr);

            //convert back to string
            String key = new String(charArr);

            //check if the key doesn't exist
            if(!mp.containsKey(key)) {
                mp.put(key, new ArrayList<>());
            }

            //add it to the map
            mp.get(key).add(val);
        }

        List<List<String>> ans = new ArrayList<>();

        // we can also do: return new ArrayList<>(mp.values());
        
        for(String key : mp.keySet()) {
            ans.add(mp.get(key));
        }

        return ans;
    }
}